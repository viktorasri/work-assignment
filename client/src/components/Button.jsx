import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ children, variant, ...props }) => {
  return (
    <button className={`button ${variant ? `button--${variant}` : ''}`} {...props}>
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  variant: PropTypes.oneOf(['primary', 'icon']),
};

export default Button;
