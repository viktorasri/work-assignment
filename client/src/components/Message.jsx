import React from 'react';
import PropTypes from 'prop-types';

const Message = ({ children }) => {
  return <div className='message message--red'>{children}</div>;
};

Message.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
};

export default Message;
