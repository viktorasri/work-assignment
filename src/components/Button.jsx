import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ children, variant, ...props }) => {
  return (
    <button className={`button ${variant ? `button--${variant}` : ''}`} {...props}>
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  variant: PropTypes.oneOf(['primary', 'secondary', 'icon']),
};

export default Button;
