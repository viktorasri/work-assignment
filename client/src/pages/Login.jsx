import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Page from '../components/Page';
import FormInput from '../components/FormInput';
import Button from '../components/Button';
import Spinner from '../components/Spinner';
import Message from '../components/Message';
import { login } from '../redux/actions/userActions';

const Login = ({ history }) => {
  const dispatch = useDispatch();

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const userLogin = useSelector((state) => state.userLogin);
  const { error, success, loading } = userLogin;

  useEffect(() => {
    if (success) history.push('/');
  }, [success, history]);

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(login(username, password));
  };

  return (
    <Page>
      <div className='row'>
        <div className='login block-center'>
          {loading && <Spinner />}
          {error && <Message>{error}</Message>}

          <div className='login__heading'>
            <h1>Log in</h1>
          </div>

          <form className='form' onSubmit={handleSubmit}>
            <FormInput
              onChange={(e) => setUsername(e.target.value)}
              value={username}
              placeholder='Enter Username'
              required
            />
            <FormInput
              onChange={(e) => setPassword(e.target.value)}
              value={password}
              type='password'
              placeholder='Enter Password'
              required
            />

            <div className='form__control d-flex justify-center'>
              <Button variant='primary' type='submit'>
                Login
              </Button>
            </div>
          </form>
        </div>
      </div>
    </Page>
  );
};

export default Login;
