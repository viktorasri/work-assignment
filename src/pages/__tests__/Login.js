import React from 'react';
import { shallow, mount } from 'enzyme';
import * as redux from 'react-redux';
import Login from '../Login';
import Button from '../../components/Button';
import Spinner from '../../components/Spinner';
import Message from '../../components/Message';

describe('Page <Login />', () => {
  const useSelectorSpy = jest.spyOn(redux, 'useSelector');
  const useDispatchSpy = jest.spyOn(redux, 'useDispatch');

  const mockHistory = {
    push: jest.fn(),
  };

  it('expect to render component for not signed in user', () => {
    useSelectorSpy.mockReturnValue({});
    expect(shallow(<Login />)).toMatchSnapshot();
  });

  it('expect to show spinner then loading', () => {
    useSelectorSpy.mockReturnValue({ loading: true });
    expect(shallow(<Login />).find(Spinner).length).toEqual(1);
  });

  it('expect to show error message on login fail', () => {
    useSelectorSpy.mockReturnValue({ error: 'unauthorized' });
    expect(shallow(<Login />).find(Message).length).toEqual(1);
  });

  it('should call history.push when user login successfully', () => {
    const mockProps = {
      history: mockHistory,
    };
    useSelectorSpy.mockReturnValue({ success: true });
    const wrapper = mount(<Login {...mockProps} />);
    expect(mockHistory.push).toHaveBeenCalledTimes(1);
  });
});
