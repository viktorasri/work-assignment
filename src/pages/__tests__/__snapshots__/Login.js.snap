// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page <Login /> expect to render component for not signed in user 1`] = `
<Page
  description="Server explorer web app"
  title="Login"
>
  <div
    className="row"
  >
    <div
      className="login block-center"
    >
      <div
        className="login__heading"
      >
        <h1>
          Log in
        </h1>
      </div>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <FormInput
          id="username"
          onChange={[Function]}
          placeholder="Enter Username"
          required={true}
          type="text"
          value=""
        />
        <FormInput
          id="password"
          onChange={[Function]}
          placeholder="Enter Password"
          required={true}
          type="password"
          value=""
        />
        <div
          className="form__control d-flex justify-center"
        >
          <Button
            id="login-btn"
            type="submit"
            variant="primary"
          >
            Login
          </Button>
        </div>
      </form>
    </div>
  </div>
</Page>
`;
